garage/np/algos/base.py

```diff
@@ -1,7 +1,91 @@
-class Algorithm:
-    pass
+from collections import deque
 
+from dowel import tabular
+import numpy as np
 
-class RLAlgorithm(Algorithm):
-    def train(self):
-        raise NotImplementedError
+from garage.misc import special, tensor_utils
+
+
+class RLAlgorithm():
+    def __init__(self,
+                 env_spec,
+                 policy,
+                 baseline,
+                 n_samples,
+                 gae_lambda=1,
+                 discount=0.99,
+                 max_path_length=500):
+        self.env_spec = env_spec
+        self.policy = policy
+        self.baseline = baseline
+        self.n_samples = n_samples
+        self.gae_lambda = gae_lambda
+        self.discount = discount
+        self.max_path_length = max_path_length
+
+        self.episode_reward_mean = deque(maxlen=100)
+
+    def process_samples(self, itr, paths):
+        """Return processed sample data based on the collected paths.
+        This method is currently used by cem and cam_es.
+        Args:
+            itr (int): Iteration number.
+            paths (list[dict]): A list of collected paths.
+        Returns:
+            dict: Processed sample data. Only {'average_return': float]}
+        """
+        baselines = []
+        returns = []
+
+        max_path_length = self.max_path_length
+
+        if hasattr(self.baseline, 'predict_n'):
+            all_path_baselines = self.baseline.predict_n(paths)
+        else:
+            all_path_baselines = [
+                self.baseline.predict(path) for path in paths
+            ]
+
+        for idx, path in enumerate(paths):
+            # baselines
+            path['baselines'] = all_path_baselines[idx]
+            baselines.append(path['baselines'])
+
+            # returns
+            path['returns'] = special.discount_cumsum(path['rewards'],
+                                                      self.discount)
+            returns.append(path['returns'])
+
+        agent_infos = [path['agent_infos'] for path in paths]
+        agent_infos = tensor_utils.stack_tensor_dict_list([
+            tensor_utils.pad_tensor_dict(p, max_path_length)
+            for p in agent_infos
+        ])
+
+        valids = [np.ones_like(path['returns']) for path in paths]
+        valids = tensor_utils.pad_tensor_n(valids, max_path_length)
+
+        average_discounted_return = (np.mean(
+            [path['returns'][0] for path in paths]))
+
+        undiscounted_returns = [sum(path['rewards']) for path in paths]
+        self.episode_reward_mean.extend(undiscounted_returns)
+
+        ent = np.sum(self.policy.distribution.entropy(agent_infos) *
+                     valids) / np.sum(valids)
+
+        samples_data = dict(average_return=np.mean(undiscounted_returns))
```

Comment:
if you only return `average_return`, why calculate all this other stuff?

Reply:
if you are going to do these calculations/logging in the base class, make sure that the inheriting classes can reuse them, rather than having to recalculate everything again